### PR TITLE
[FIXED JENKINS-15331] Retry failed file delete operations

### PR DESCRIPTION
**Pull request by pjdarton**
This changes Jenkins' strategy when doing recursive deletes so that it no longer stops iterating through a directory when it finds a file it can't delete - it'll try to delete what it can before throwing the exception.
It also changes the default number of attempts that Jenkins makes when doing a recursive delete from 2 attempts to 3 attempts.
This combination results in a dramatic improvement in deletion reliability on Windows-based job executors.

This behavior is configurable (on a per-JVM basis), albeit crudely - like the existing hudson.Util.noSymlink & hudson.Util.symlinkEscapeHatch flags, these can be configured using -D options when starting the Jenkins or slave services.

-Dhudson.Util.deletionMax=3
Sets the number of times Jenkins will try to delete a file before giving up.
Defaults to 3 attempts.

-Dhudson.Util.deletionRetryWait=100
Sets the time (in milliseconds) for which Jenkins will pause between delete attempts.
Defaults to 100ms between each attempt.
If set to a negative number, it will delay for a linearly increasing amount between attempts.

-Dhudson.Util.performGCOnFailedDelete=false
If set to true, this tells Jenkins to run the Garbage Collector between attempts (just in case the process holding a file open is an un-garbage-collected Java filehandle within Jenkins itself).
It defaults to false.

(Note: This was reworked to patch the Jenkins master branch as of 15:30 GMT Aug 17 2015, and superceeds pull request 1209)

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
I think the change is reasonable even if it does not guarantee the deletion success. :+1: for merge

**Comment by jtnord** [comment]
> This combination results in a dramatic improvement in deletion reliability on Windows-based job executors.

Are there issues for this, otherwise we are just masking issues rather than fixing them?

**Comment by pjdarton** [comment]
@jtnord https://issues.jenkins-ci.org/browse/JENKINS-15331 details the solution, but there are many other issues showing the way the problem manifests (most typically during the "delete workspace" phase of a build), several of which are either "duplicates of 15331" or "related to 15331".
Basically, you can't *reliably* delete (or move, or open for writing) files/directories on Windows - the filesystem is inherently non-deterministic because of (arguably poor) design decisions made by Microsoft, which manifests as a lack of reliability.  Code running on Windows needs to cope with that in order to avoid passing that "unreliable" behavior back to the user.

So, yes, technically we *are* "just masking issues rather than fixing them", but the "proper" fix would be either "don't use Windows!" (impractical), or have Microsoft implement a fix (unlikely - they don't acknowledge that this is a problem and they have been ignoring this for over a decade now); this is a pragmatic solution to the problem.

**Comment by jtnord** [comment]
@pjdarton What I was asking is is there issues where this occurs so that we can look at fixing the root issues rather than masking the issue (and potentially having falky builds).  

If a full GC solves the issue then a plugin or core has not closed the file after using it - so this should be an easy fix if we can narrow this down to the plugins in use.  
External tools are a little more complicated - some behave well, and some do not.  We could use some hooks to find out what that tool is (given we know the file path) and file upstream issues (if that tool is designed to live outside the build).  The other suspect area would be workspace browsing or SCMs which require the workspace for polling.

**Comment by pjdarton** [comment]
The root issue really is the Windows OS.  It's not a coding fault within Jenkins.  It's not caused by Jenkins code holding files open, it's caused by other processes on the Windows slave interfering with the files which "belong" to the Jenkins slave, e.g. by using the Windows "filesystem activity hook" (as used by anti-virus, Windows Search etc) to detect the addition (or usage) of files on the filesystem, and then opening the files that Jenkins is about to remove (e.g. in order to inspect them), which then *sometimes* causes the deletion to fail (because another process briefly has the file open).
This is a "well known" problem on Windows.
e.g. if you use Windows Explorer to delete a deeply-nested directory hierarchy, it'll often leave you with some empty folders that you have to delete again.  Similarly, the Windows CMD window "RD /S /Q myNestedDirecoty" command will sometimes state "directory is not empty" and require you to repeat the command.
i.e. this is not a problem within Jenkins, or its code, it's a problem on Windows.

The solution, on Windows, is to issue the deletion request multiple times.  Usually, the rogue process (e.g. Windows Search Service) that locked the file open won't keep it open very long, so an immediate retry will succeed.
If the underlying problem is that Jenkins itself has the file open (e.g. the RTC plugin has a bug whereby it can attempt to reinstall the RTC jars while the RTC plugin is using them) then no amount of retrying will fix that, at which point the build will fail with an error message (albeit taking slightly longer to do so).
Similarly, if an external process has a file (locked) open then the build will still fail (and report why).  What this code does is handle the situation where files are *briefly* locked open while the delete is ongoing (which is typical behavior of the Windows Search Service, and most Anti-Virus software, but this malware-like-behavior is not limited to just those two).
i.e. The code enhancement I've done does not mask *real* persistent failures, only transient ones caused by the (poor) Windows filesystem behavior.

Re: GC
I would agree that, if a full GC "solved" an issue, then that *would* be just masking the issue, but this is why the GC behavior is turned off by default.  This functionality is useful when encountering buggy plugins (typically written by people who are 100% unix so they never encounter file deletion problems), and allows you to temporarily work around such bugs until a proper fix can be implemented.
FYI Apache Ant uses the same GC technique in its deletion code.  I was persuaded (see https://github.com/jenkinsci/jenkins/pull/615) that it should be disabled by default in Jenkins, and the javadoc for the GC configuration flag came from that discussion.
i.e. I'm not recommending that the GC functionality is enabled by default, and I wouldn't describe it as a "cure", but it can be an effective "workaround", allowing people to continue to work while a bugfix is done.

In summary:
 - The deletion retry mechanism is a workaround for a "behavioral characteristic" (aka a bug) in Windows, and (in my experience) is essential (without it, one sometimes gets red builds which aren't the developer's fault).
 - The GC capability is less essential but is *sometimes* very useful when running code that has yet to be fixed to also run on Windows.  It's a very useful capability to have available "in case of emergency", and using this can *really* help with tracking down the cause of locked files (as, if this fixes it, you know that your code is relying on the GC to close files that it should have closed itself).

**Comment by jtnord** [comment]
@pjdarton thanks for the detailed explanation

**Comment by oleg-nenashev** [comment]
I think we can integrate it into the master branch and see how it goes. :+1: from me.
@jtnord WDYT?

We could convert it to the experimental feature as a plan B

**Comment by pjdarton** [comment]
I would very much like this integrated into the master branch.

Re:"experimental feature", it can be configured that way if the defaults
don't suit, just as with other settings in the Util class, however I
believe that the chosen defaults should work well as-is.  I can vouch that
those settings work fine (for me and my colleagues) on both Windows and
Linux.
FYI The text for the wiki page
https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties
could be copy/pasted straight from the Javadoc in the code - I wrote those
comments with the aim of eliminating the need for people to ask me what
those options did ;)

File: core/src/main/java/hudson/Util.java [hunk 1]
```diff
@@ -1498,4 +1653,54 @@ public static Properties loadProperties(@Nonnull String properties) throws IOExc
...
+     * failing because Jenkins is unable to delete files, that this failure is
+     * because Jenkins itself has those files locked "open", and even then it
+     * should only be used on slaves with relatively few executors (because the
+     * garbage collection can impact the performance of all job executors on
+     * that slave).<br/>
+     * i.e. Setting this flag is a act of last resort - it is <em>not</em>
+     * recommended, and should not be used on the main Jenkins server
+     * unless you can tolerate the performance impact.
+     */
+    public static boolean GC_AFTER_FAILED_DELETE = Boolean.getBoolean(Util.class.getName() + ".performGCOnFailedDelete");
```
Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: private please - or package if you need it for tests.

File: core/src/main/java/hudson/Util.java [hunk 2]
```diff
@@ -1498,4 +1653,54 @@ public static Properties loadProperties(@Nonnull String properties) throws IOExc
...
+     * <p>
+     * e.g. if some of the child directories are big, it might take long enough
+     * to delete that it allows others to create new files in the directory we
+     * are trying to empty, causing problems like JENKINS-10113.
+     * Or, if we're on Windows, then deletes can fail for transient reasons
+     * regardless of external activity; see JENKINS-15331.
+     * Whatever the reason, this allows us to do multiple attempts before we
+     * give up, thus improving build reliability.
+     */
+    public static int DELETION_MAX = Math.max(1, Integer.getInteger(Util.class.getName() + ".deletionMax", 3).intValue());
```
Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: private please or package if you need it for tests.

File: core/src/main/java/hudson/Util.java [hunk 3]
```diff
@@ -1498,4 +1653,54 @@ public static Properties loadProperties(@Nonnull String properties) throws IOExc
...
+    /**
+     * The time (in milliseconds) that we will wait between attempts to
+     * delete files when retrying.<br>
+     * This has no effect unless {@link #DELETION_MAX} is non-zero.
+     * <p>
+     * If zero, we will not delay between attempts.<br>
+     * If negative, we will wait an (linearly) increasing multiple of this value
+     * between attempts.
+     */
+    public static int WAIT_BETWEEN_DELETION_RETRIES = Integer.getInteger(Util.class.getName() + ".deletionRetryWait", 100).intValue();
```
Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: private please or package if you need it for tests.

**Comment by jtnord** [comment]
I'm ok on it going in if we get rid of the newly exposed public state.
I see no reason why this needs to be public.  You can change private state from the script console - and I do not want more public fields causing issues in the future.

**Comment by daniel-beck** [comment]
Should probably also be `@Restricted(NoExternalUse.class)` to be safer.

**Comment by pjdarton** [comment]
@daniel-beck I'm unfamiliar with that particular annotation.  Can you
elaborate?
e.g. Is this a mark-up to be applied to the (now package level access)
fields to indicate that they are not to be used by anything other than the
unit-test class?

**Comment by daniel-beck** [comment]
@pjdarton Exactly. Sorry I missed the notification.

Something appears to have gone really wrong with this PR. There are tons of unrelated commits. Could you please fix this?

**Comment by pjdarton** [comment]
@daniel-beck My skills with the git command-line are somewhat limited (as demonstrated by the presence of "tons of unrelated commits"...), so I've fixes this mess the only way I know - I have wiped and re-created the branch.  That automatically closed this pull request, so I've had to create a new one - https://github.com/jenkinsci/jenkins/pull/2026

I hope that this new (and clean!) pull request will meet all your requirements and we can (finally) get this functionality merged into the Jenkins code.
